File: python/913.py
"""
Project Euler Problem 913
"""
import math

def gcd(a, b):
    while b:
        a, b = b, a % b
    return a

def get_prime_factors_small(n):
    """
    Returns a dictionary of prime factors for small n (n <= 10^8).
    """
    factors = {}
    d = 2
    temp = n
    while d * d <= temp:
        while temp % d == 0:
            factors[d] = factors.get(d, 0) + 1
            temp //= d
        d += 1
    if temp > 1:
        factors[temp] = factors.get(temp, 0) + 1
    return factors

def get_prime_factors_of_L(n, m):
    """
    Factors L = n^4 * m^4 - 1 for the large problem,
    or L = n * m - 1 for the small problem if needed.
    Wait, for small problem L = nm - 1.
    For large problem L = (nm)^4 - 1.
    
    This function factors X^4 - 1 where X = nm.
    L = (X-1)(X+1)(X^2+1).
    Since X <= 10000, X^2+1 <= 10^8 + 1.
    We can factor each part using trial division.
    """
    X = n * m
    # Factors of X - 1
    factors = get_prime_factors_small(X - 1)
    
    # Factors of X + 1
    f2 = get_prime_factors_small(X + 1)
    for p, count in f2.items():
        factors[p] = factors.get(p, 0) + count

    # Factors of X^2 + 1
    f3 = get_prime_factors_small(X * X + 1)
    for p, count in f3.items():
        factors[p] = factors.get(p, 0) + count

    return factors

def pow_mod(a, b, m):
    res = 1
    a %= m
    while b > 0:
        if b % 2 == 1:
            res = (res * a) % m
        a = (a * a) % m
        b //= 2
    return res

def get_order_mod_pk(a, p, k):
    """
    Computes order of a modulo p^k.
    Precondition: gcd(a, p) = 1.
    """
    # First compute order mod p
    # order divides p-1
    # Factor p-1
    if p == 2:
        # For p=2, order is usually power of 2
        # But group structure of (Z/2^k Z)* is not cyclic for k>=3.
        # However, we just need the smallest d such that a^d = 1 mod 2^k.
        # Just check powers of 2?
        # a must be odd.
        # Order divides phi(2^k) = 2^(k-1).
        # So order is a power of 2.
        order = 1
        curr = a % (2**k)
        if curr == 1:
            return 1
        # Try 2, 4, 8...
        for i in range(1, k+1): # max power is k-1? 2^(k-1)
            order *= 2
            curr = (curr * curr) % (2**k)
            if curr == 1:
                return order
        return order

    # For odd p
    p_minus_1 = p - 1
    factors = get_prime_factors_small(p_minus_1)
    order = p_minus_1
    for q, count in factors.items():
        # Try to divide order by q as much as possible
        while order % q == 0:
            temp_order = order // q
            if pow_mod(a, temp_order, p) == 1:
                order = temp_order
            else:
                break

    # Now lift to p^k
    # If a^order = 1 + c*p^v, then order mod p^(v+1) is order*p...
    # We can just iterate.
    current_mod = p
    current_val = pow_mod(a, order, current_mod) # This is 1 mod p

    # We need order mod p^k.
    # We know order mod p is `order`.
    # Let order_j be order mod p^j.
    # order_{j+1} is either order_j or order_j * p.
    # Actually, generally, if ord_p(a) = d, then ord_{p^k}(a) = d * p^v for some v.
    # We can just multiply by p until it works.

    total_mod = p**k
    curr_order = order
    while True:
        if pow_mod(a, curr_order, total_mod) == 1:
            return curr_order
        curr_order *= p

def count_cycles(n, m, is_large=False):
    """
    Counts cycles for transformation of n x m matrix.
    If is_large is True, arguments are n, m but we solve for n^4, m^4.
    Otherwise we solve for n, m.
    """
    if is_large:
        # Solving for N=n^4, M=m^4
        # L = N*M - 1 = n^4 m^4 - 1
        # Multiplier A = N = n^4
        factors_L = get_prime_factors_of_L(n, m) # Factors of (nm)^4 - 1
        multiplier = pow(n, 4)
    else:
        # Solving for N=n, M=m
        # L = n*m - 1
        # Multiplier A = n
        factors_L = get_prime_factors_small(n * m - 1)
        multiplier = n

    # Precompute order of multiplier for each maximal prime power in L
    # p^e || L
    prime_power_orders = {} # p -> (e, order_mod_p_e, phi_p_e)
    primes = sorted(factors_L.keys())

    for p in primes:
        e = factors_L[p]
        pk = p**e
        ord_val = get_order_mod_pk(multiplier, p, e)
        phi_val = (p**(e-1)) * (p-1)
        prime_power_orders[p] = (e, ord_val, phi_val)

    # Now DFS to sum phi(d)/ord_d(multiplier)
    # We build d prime by prime
    # For each prime p with exponent e in L (so p^e || L):
    # divisor d can have p^j where 0 <= j <= e.
    # phi(d) is multiplicative.
    # ord_d(mult) = lcm(ord_{p^j}(mult)) over p^j || d.
    # Note ord_{p^j} is needed.
    # prime_power_orders stores ord for p^e.
    # We need ord for p^j for j < e.
    # It is efficient to compute them on demand or precompute all.
    # Since e is small (usually 1, maybe up to 13 for 2^13 < 10000?),
    # we can precompute for all j.

    # Better precompute structure: p -> list of (phi(p^j), ord(p^j)) for j=0..e
    precomputed = []
    for p in primes:
        e = factors_L[p]
        entry = []
        # j=0 => d=1, phi=1, ord=1
        entry.append((1, 1))

        current_p_pow = 1
        current_phi = 1
        # We need ord mod p^j.
        # We can compute it from scratch or derive from higher order?
        # Safer to compute from scratch or incrementally.

        # Incremental computation
        # ord mod p
        ord_p = get_order_mod_pk(multiplier, p, 1)
        curr_ord = ord_p
        current_p_pow = p

        # For j=1
        entry.append((p-1, curr_ord))

        for j in range(2, e+1):
            current_p_pow *= p
            # Check if curr_ord works for new power
            if pow_mod(multiplier, curr_ord, current_p_pow) != 1:
                curr_ord *= p
            entry.append((p**(j-1) * (p-1), curr_ord))

        precomputed.append(entry)

    # DFS
    # State: (index in primes, current_lcm_order, current_phi_product)

    total_sum = 0

    stack = [(0, 1, 1)] # index, lcm_order, phi_prod

    while stack:
        idx, lcm_ord, phi_prod = stack.pop()

        if idx == len(primes):
            total_sum += phi_prod // lcm_ord
            continue

        # Try all exponents for primes[idx]
        # entries is list of (phi, ord) for exponents 0..e
        entries = precomputed[idx]
        for phi_part, ord_part in entries:
            # new_lcm = lcm(lcm_ord, ord_part)
            # lcm(a, b) = (a*b)//gcd(a, b)
            g = gcd(lcm_ord, ord_part)
            new_lcm = (lcm_ord * ord_part) // g

            stack.append((idx + 1, new_lcm, phi_prod * phi_part))

    # The sum counts cycles in {0, ..., L-1}.
    # This includes the cycle {0}.
    # The cycles we want are in {1, ..., L-1}.
    # So we have (total_sum - 1) cycles in {1, ..., nm-2}.
    # And we add fixed points 0 and nm-1 (indices).
    # Wait, the formula S = nm - (cycles in 1..nm-2) - 2 ?
    # From my thought process:
    # Cycles_in_range = total_sum - 1.
    # Total_Cycles_Permutation = Cycles_in_range + 2.
    # S = Size - Total_Cycles_Permutation
    # S = Size - (total_sum - 1 + 2) = Size - total_sum - 1.
    
    return total_sum

def S(n, m, is_large=False):
    if is_large:
        size = (n*m)**4
    else:
        size = n*m

    if size <= 2: return 0 # Trivial cases
    
    # corner case: if size-1 == 1, loop doesn't work well?
    # nm=2 -> size-1=1. L=1. Divisors d=1.
    # sum = phi(1)/ord(1) = 1.
    # S = 2 - 1 - 1 = 0. Correct.

    cycle_term = count_cycles(n, m, is_large)
    return size - cycle_term - 1

def solve_small():
    total = 0
    for n in range(2, 101):
        for m in range(n, 101):
            val = S(n, m, is_large=False)
            total += val
    return total

def solve_large():
    total = 0
    for n in range(2, 101):
        for m in range(n, 101):
            val = S(n, m, is_large=True)
            total += val
    return total

def solve():
    return solve_large()

if __name__ == "__main__":
    import sys
    if len(sys.argv) > 1 and sys.argv[1] == "small":
        print(solve_small())
    else:
        print(solve())

File: c/913.c
/* Project Euler Problem 913 - Matrix Shuffles
 * For (n,m) with 2<=n<=m<=100, compute S(n,m) for the n^4,m^4 case.
 * S = (nm)^4 - cycle_count - 1
 * cycle_count = sum phi(d)/ord_d(multiplier) over divisors of L=(nm)^4-1.
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef long long ll;
typedef unsigned long long ull;
typedef __int128 lll;

ll gcd_ll(ll a, ll b) {
    if (a < 0) a = -a;
    if (b < 0) b = -b;
    while (b) { ll t = b; b = a % b; a = t; }
    return a;
}

ll lcm_ll(ll a, ll b) {
    return a / gcd_ll(a, b) * b;
}

/* Trial division factorization of n into primes */
typedef struct { ll prime; int exp; } Factor;

int factorize(ll n, Factor *factors) {
    int cnt = 0;
    for (ll d = 2; d * d <= n; d++) {
        if (n % d == 0) {
            factors[cnt].prime = d;
            factors[cnt].exp = 0;
            while (n % d == 0) {
                factors[cnt].exp++;
                n /= d;
            }
            cnt++;
        }
    }
    if (n > 1) {
        factors[cnt].prime = n;
        factors[cnt].exp = 1;
        cnt++;
    }
    return cnt;
}

ll pow_mod_ll(ll a, ll b, ll m) {
    ll res = 1;
    a %= m;
    if (a < 0) a += m;
    while (b > 0) {
        if (b & 1) res = (lll)res * a % m;
        a = (lll)a * a % m;
        b >>= 1;
    }
    return res;
}

/* Compute order of a mod p^k */
ll get_order_mod_pk(ll a, ll p, int k) {
    ll pk = 1;
    for (int i = 0; i < k; i++) pk *= p;

    if (p == 2) {
        ll order = 1;
        ll curr = a % pk;
        if (curr < 0) curr += pk;
        if (curr == 1) return 1;
        for (int i = 1; i <= k; i++) {
            order *= 2;
            curr = (lll)curr * curr % pk;
            if (curr == 1) return order;
        }
        return order;
    }

    /* For odd p: find order mod p first */
    ll p_minus_1 = p - 1;
    Factor pf[64];
    int npf = factorize(p_minus_1, pf);

    ll order = p_minus_1;
    for (int i = 0; i < npf; i++) {
        ll q = pf[i].prime;
        while (order % q == 0) {
            ll temp_order = order / q;
            if (pow_mod_ll(a, temp_order, p) == 1)
                order = temp_order;
            else
                break;
        }
    }

    /* Lift to p^k */
    ll total_mod = pk;
    ll curr_order = order;
    while (1) {
        if (pow_mod_ll(a, curr_order, total_mod) == 1)
            return curr_order;
        curr_order *= p;
    }
}

/* Precomputed data per prime factor */
typedef struct {
    int num_entries; /* 0..e entries */
    ll phi_val[64];  /* phi(p^j) for j=0..e */
    ll ord_val[64];  /* ord mod p^j for j=0..e */
} PrimeData;

ll count_cycles(int n, int m) {
    ll X = (ll)n * m;
    /* L = X^4 - 1 = (X-1)(X+1)(X^2+1) */
    /* Factor each part */
    Factor all_factors[256];
    int total_factors = 0;

    ll parts[3] = { X - 1, X + 1, X * X + 1 };
    Factor pf[64];

    /* Merge all prime factors */
    typedef struct { ll prime; int exp; } PF;
    PF merged[256];
    int nmerged = 0;

    for (int pi = 0; pi < 3; pi++) {
        int nf = factorize(parts[pi], pf);
        for (int i = 0; i < nf; i++) {
            int found = 0;
            for (int j = 0; j < nmerged; j++) {
                if (merged[j].prime == pf[i].prime) {
                    merged[j].exp += pf[i].exp;
                    found = 1;
                    break;
                }
            }
            if (!found) {
                merged[nmerged].prime = pf[i].prime;
                merged[nmerged].exp = pf[i].exp;
                nmerged++;
            }
        }
    }

    ll multiplier_raw = (ll)n * n * n * n; /* n^4 */

    /* Precompute data for each prime */
    PrimeData pd[256];
    for (int i = 0; i < nmerged; i++) {
        ll p = merged[i].prime;
        int e = merged[i].exp;
        pd[i].num_entries = e + 1;

        /* j=0: phi=1, ord=1 */
        pd[i].phi_val[0] = 1;
        pd[i].ord_val[0] = 1;

        /* j=1 */
        ll ord_p = get_order_mod_pk(multiplier_raw, p, 1);
        ll curr_ord = ord_p;
        pd[i].phi_val[1] = p - 1;
        pd[i].ord_val[1] = curr_ord;

        ll current_p_pow = p;
        for (int j = 2; j <= e; j++) {
            current_p_pow *= p;
            if (pow_mod_ll(multiplier_raw, curr_ord, current_p_pow) != 1)
                curr_ord *= p;
            ll phi_j = 1;
            ll pp = 1;
            for (int t = 0; t < j - 1; t++) pp *= p;
            phi_j = pp * (p - 1);
            pd[i].phi_val[j] = phi_j;
            pd[i].ord_val[j] = curr_ord;
        }
    }

    /* DFS to sum phi(d)/ord_d(mult) */
    typedef struct { int idx; ll lcm_ord; ll phi_prod; } StackEntry;
    StackEntry stack[1000000];
    int sp = 0;
    stack[sp++] = (StackEntry){0, 1, 1};

    ll total_sum = 0;

    while (sp > 0) {
        StackEntry se = stack[--sp];
        if (se.idx == nmerged) {
            total_sum += se.phi_prod / se.lcm_ord;
            continue;
        }
        int idx = se.idx;
        for (int j = 0; j < pd[idx].num_entries; j++) {
            ll phi_part = pd[idx].phi_val[j];
            ll ord_part = pd[idx].ord_val[j];
            ll g = gcd_ll(se.lcm_ord, ord_part);
            ll new_lcm = se.lcm_ord / g * ord_part;
            stack[sp++] = (StackEntry){idx + 1, new_lcm, se.phi_prod * phi_part};
        }
    }

    return total_sum;
}

int main(void) {
    ll total = 0;
    for (int n = 2; n <= 100; n++) {
        for (int m = n; m <= 100; m++) {
            ll size = 1;
            for (int i = 0; i < 4; i++) size *= (ll)n * m;
            if (size <= 2) continue;
            ll cycle_term = count_cycles(n, m);
            ll val = size - cycle_term - 1;
            total += val;
        }
    }
    printf("%lld\n", total);
    return 0;
}
